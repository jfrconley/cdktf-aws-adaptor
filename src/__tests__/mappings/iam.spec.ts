import { IamAccessKey } from "@cdktf/provider-aws/lib/iam-access-key/index.js";
import { IamGroupPolicyAttachment } from "@cdktf/provider-aws/lib/iam-group-policy-attachment/index.js";
import { IamPolicy } from "@cdktf/provider-aws/lib/iam-policy/index.js";
import { IamRolePolicyAttachment } from "@cdktf/provider-aws/lib/iam-role-policy-attachment/index.js";
import { IamRole } from "@cdktf/provider-aws/lib/iam-role/index.js";
import { IamUserPolicyAttachment } from "@cdktf/provider-aws/lib/iam-user-policy-attachment/index.js";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket/index.js";
import { CfnAccessKey, CfnPolicy, CfnRole } from "aws-cdk-lib/aws-iam";
import { Testing } from "cdktf";
import { resolve } from "cdktf/lib/_tokens.js";
import { setupJest } from "cdktf/lib/testing/adapters/jest.js";
import { itShouldMapCfnElementToTerraformResource, synthesizeElementAndTestStability } from "../helpers.js";
import { describe, it } from "vitest";

setupJest();
describe("IAM Mappings", () => {
    it("should map AWS::IAM::Role", () => {
        const {
            stack: testStack,
            resource: iamRole,
        } = synthesizeElementAndTestStability(
            CfnRole,
            {
                path: "packages/cdktf-adaptor/src/mappings/services/iam.ts",
                description: "IamRole",
                tags: [
                    {
                        key: "Name",
                        value: "test-role",
                    },
                ],
                managedPolicyArns: [
                    "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role",
                ],
                maxSessionDuration: 3600,
                roleName: "test-role",
                assumeRolePolicyDocument: {
                    Version: "2012-10-17",
                    Statement: [
                        {
                            Effect: "Allow",
                            Principal: {
                                Service: "ec2.amazonaws.com",
                            },
                        },
                    ],
                },
                policies: [
                    {
                        policyName: "test-policy",
                        policyDocument: {
                            Version: "2012-10-17",
                            Statement: [
                                {
                                    Effect: "Allow",
                                },
                            ],
                        },
                    },
                ],
                permissionsBoundary: "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role",
            },
            IamRole,
            {
                path: "packages/cdktf-adaptor/src/mappings/services/iam.ts",
                name: "test-role",
                assumeRolePolicy:
                    "$\{jsonencode(\{\"Version\" = \"2012-10-17\", \"Statement\" = [{\"Effect\" = \"Allow\", \"Principal\" = {\"Service\" = \"ec2.amazonaws.com\"}}]})\}",
                description: "IamRole",
                inlinePolicy: [
                    {
                        name: "test-policy",
                        policy:
                            "$\{jsonencode(\{\"Version\" = \"2012-10-17\", \"Statement\" = [{\"Effect\" = \"Allow\"}]})\}",
                    },
                ],
                permissionsBoundary: "arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role",
                tags: {
                    Name: "test-role",
                },
                maxSessionDuration: 3600,
            },
        );

        new S3Bucket(testStack, "test-bucket", {
            bucket: iamRole.arn,
        });

        const synth = Testing.synth(testStack);

        expect(synth).toHaveResourceWithProperties(S3Bucket, {
            depends_on: [
                "time_sleep.resource_resource-wait_D049D385",
                "aws_iam_role_policy_attachment.resource_resource-managed-policy-0_D5E0A017",
            ],
        });
    });

    it("should map AWS::IAM::Policy", () => {
        const {
            synth,
            stack,
            resource,
        } = synthesizeElementAndTestStability(
            CfnPolicy,
            {
                policyName: "test-policy",
                policyDocument: {
                    Version: "2012-10-17",
                    Statement: [
                        {
                            Effect: "Allow",
                        },
                    ],
                },
                groups: ["test-group"],
                roles: ["test-role"],
                users: ["test-user"],
            },
            IamPolicy,
            {
                name: "test-policy",
                policy: "$\{jsonencode(\{\"Version\" = \"2012-10-17\", \"Statement\" = [{\"Effect\" = \"Allow\"}]})\}",
            },
        );

        expect(synth).toHaveResourceWithProperties(IamRolePolicyAttachment, {
            policy_arn: resolve(stack, resource.arn),
            role: "test-role",
        });

        expect(synth).toHaveResourceWithProperties(IamGroupPolicyAttachment, {
            policy_arn: resolve(stack, resource.arn),
            group: "test-group",
        });

        expect(synth).toHaveResourceWithProperties(IamUserPolicyAttachment, {
            policy_arn: resolve(stack, resource.arn),
            user: "test-user",
        });
    });

    itShouldMapCfnElementToTerraformResource(
        CfnAccessKey,
        {
            status: "Active",
            userName: "test-user",
        },
        IamAccessKey,
        {
            status: "Active",
            user: "test-user",
        },
    );
});
