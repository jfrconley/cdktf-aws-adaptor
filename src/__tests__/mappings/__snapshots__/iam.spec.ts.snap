// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IAM Mappings > should map AWS::IAM::Policy 1`] = `
"{
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_iam_group_policy_attachment\\": {
      \\"resource_resource_group0_AB0AE204\\": {
        \\"group\\": \\"test-group\\",
        \\"policy_arn\\": \\"\${aws_iam_policy.resource_22C949BF.arn}\\"
      }
    },
    \\"aws_iam_policy\\": {
      \\"resource_22C949BF\\": {
        \\"name\\": \\"test-policy\\",
        \\"policy\\": \\"\${jsonencode({\\\\\\"Version\\\\\\" = \\\\\\"2012-10-17\\\\\\", \\\\\\"Statement\\\\\\" = [{\\\\\\"Effect\\\\\\" = \\\\\\"Allow\\\\\\"}]})}\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"resource_resource_role0_CD7751A8\\": {
        \\"policy_arn\\": \\"\${aws_iam_policy.resource_22C949BF.arn}\\",
        \\"role\\": \\"test-role\\"
      }
    },
    \\"aws_iam_user_policy_attachment\\": {
      \\"resource_resource_user0_1355940A\\": {
        \\"policy_arn\\": \\"\${aws_iam_policy.resource_22C949BF.arn}\\",
        \\"user\\": \\"test-user\\"
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.40.0\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${aws_iam_policy.resource_22C949BF.id}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`IAM Mappings > should map AWS::IAM::Role 1`] = `
"{
  \\"data\\": {
    \\"aws_iam_role\\": {
      \\"resource_resource-data_E9A5FDBC\\": {
        \\"depends_on\\": [
          \\"aws_iam_role_policy_attachment.resource_resource-managed-policy-0_D5E0A017\\"
        ],
        \\"name\\": \\"\${aws_iam_role.resource_22C949BF.name}\\"
      }
    }
  },
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_iam_role\\": {
      \\"resource_22C949BF\\": {
        \\"assume_role_policy\\": \\"\${jsonencode({\\\\\\"Version\\\\\\" = \\\\\\"2012-10-17\\\\\\", \\\\\\"Statement\\\\\\" = [{\\\\\\"Effect\\\\\\" = \\\\\\"Allow\\\\\\", \\\\\\"Principal\\\\\\" = {\\\\\\"Service\\\\\\" = \\\\\\"ec2.amazonaws.com\\\\\\"}}]})}\\",
        \\"description\\": \\"IamRole\\",
        \\"inline_policy\\": [
          {
            \\"name\\": \\"test-policy\\",
            \\"policy\\": \\"\${jsonencode({\\\\\\"Version\\\\\\" = \\\\\\"2012-10-17\\\\\\", \\\\\\"Statement\\\\\\" = [{\\\\\\"Effect\\\\\\" = \\\\\\"Allow\\\\\\"}]})}\\"
          }
        ],
        \\"max_session_duration\\": 3600,
        \\"name\\": \\"test-role\\",
        \\"path\\": \\"packages/cdktf-adaptor/src/mappings/services/iam.ts\\",
        \\"permissions_boundary\\": \\"arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role\\",
        \\"tags\\": {
          \\"Name\\": \\"test-role\\"
        }
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"resource_resource-managed-policy-0_D5E0A017\\": {
        \\"policy_arn\\": \\"arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role\\",
        \\"role\\": \\"\${aws_iam_role.resource_22C949BF.name}\\"
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.40.0\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${join(\\\\\\"\\\\\\", [data.aws_iam_role.resource_resource-data_E9A5FDBC.arn, \\\\\\",\\\\\\", data.aws_iam_role.resource_resource-data_E9A5FDBC.id])}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;
